internal/controller/api_to_config.go

```diff
@@ -255,111 +255,169 @@ func passwordForNeighbor(n v1beta1.Neighbor, passwordSecrets map[string]corev1.S
+		res.PrefixesV6 = sets.List(prefixesForFamily[ipfamily.IPv6])
 	}
-	res := frr.AllowedOut{
-		PrefixesV4: sortMap(advsV4),
-		PrefixesV6: sortMap(advsV6),
+
+	for _, ipFamily := range neighborIPFamilies {
+		var err error
+		res.LocalPrefPrefixesModifiers, err = prefixesWithLocalPrefToFRR(res.LocalPrefPrefixesModifiers, neighbor, toAdvertise, ipFamily, prefixesForFamily[ipFamily])
+		if err != nil {
+			return frr.AllowedOut{}, fmt.Errorf("failed to process local pref for neighbor %s, err: %w", neighbor.Name, err)
+		}
+		res.CommunityPrefixesModifiers, err = prefixesWithCommunityToFRR(res.CommunityPrefixesModifiers, neighbor, toAdvertise, ipFamily, prefixesForFamily[ipFamily])
+		if err != nil {
+			return frr.AllowedOut{}, fmt.Errorf("failed to process local pref for neighbor %s, err: %w", neighbor.Name, err)
+		}
 	}
+
 	return res, nil
 }
 
-// prefixesToMap returns two maps of prefix->OutgoingFilter (ie family, advertisement, communities), one for each family.
-// The ipv4Prefixes and ipv6Prefixes represent the "global" allowed prefixes which are the prefixes defined on the router.
-func prefixesToMap(toAdvertise v1beta1.Advertise, ipv4Prefixes, ipv6Prefixes []string) (map[string]*frr.OutgoingFilter, map[string]*frr.OutgoingFilter) {
-	resV4 := map[string]*frr.OutgoingFilter{}
-	resV6 := map[string]*frr.OutgoingFilter{}
-	if toAdvertise.Allowed.Mode == v1beta1.AllowAll {
-		for _, p := range ipv4Prefixes {
-			resV4[p] = &frr.OutgoingFilter{Prefix: p, IPFamily: ipfamily.IPv4}
+func prefixesWithLocalPrefToFRR(toAdd map[string]frr.LocalPrefPrefixList, neighbor *frr.NeighborConfig, toAdvertise v1beta1.Advertise, ipFamily ipfamily.Family, routerPrefixes sets.Set[string]) (map[string]frr.LocalPrefPrefixList, error) {
+	frrFamily := frrIPFamily(ipFamily)
+	for _, prefixes := range toAdvertise.PrefixesWithLocalPref {
+		key := localPrefPrefixListKey(prefixes.LocalPref, frrFamily)
+
+		if _, ok := toAdd[key]; ok {
+			return nil, fmt.Errorf("local preference %d is already defined", prefixes.LocalPref)
 		}
-		for _, p := range ipv6Prefixes {
-			resV6[p] = &frr.OutgoingFilter{Prefix: p, IPFamily: ipfamily.IPv6}
+
+		localPrefPrefixList := frr.LocalPrefPrefixList{
+			PrefixList: frr.PrefixList{
+				Name:     localPrefPrefixListName(neighbor.ID(), prefixes.LocalPref, frrFamily),
+				IPFamily: frrFamily,
+				Prefixes: sets.New[string](),
+			},
+			LocalPref: prefixes.LocalPref,
 		}
-		return resV4, resV6
+
+		ipfamilyPrefixes := ipfamily.FilterPrefixes(prefixes.Prefixes, ipFamily)
+		for _, prefix := range ipfamilyPrefixes {
+			if !routerPrefixes.Has(prefix) {
+				return nil, fmt.Errorf("localPref %d associated to non existing prefix %s", prefixes.LocalPref, prefix)
+			}
+			if localPrefPrefixList.Prefixes.Has(prefix) {
+				return nil, fmt.Errorf("prefix %s is already defined for local preference %d", prefix, prefixes.LocalPref)
+			}
+			if existing, ok := toAdd[prefix]; ok && existing.LocalPref != prefixes.LocalPref {
```

Comment:
I might be confused, but I think `toAdd[prefix]` will never be hit? seems like only keys in a different format get in this map (it's very probable I'm misinterpreting, because I assume we already have tests for this)
maybe this is already caught by validateOutgoingPrefixes and can be removed here?

Reply:
right, we are already validating it outside. Removing